virttest/utils_linux_modules.py

```diff
@@ -0,0 +1,332 @@
+                    value = str(items[-1])
+                elif key == 'depends':
+                    value = items[1].split(',')
+                elif key == 'parm':
+                    param_dic = {'type': None}
+                    param_dic['name'] = items[1].split(':')[0]
+                    param_type = re.search(r"\((\w+)\)", items[-1])
+                    if param_type is not None:
+                        param_dic['type'] = param_type.group(1)
+                    param_list.append(param_dic)
+            if value:
+                modinfo_dic[key] = value
+        if param_list:
+            modinfo_dic['params'] = param_list
+    return modinfo_dic
+
+
+def get_submodules(module_name):
+    """
+    Get all submodules of the module.
+
+    :param module_name: Name of module to search for
+    :type module_name: str
+    :return: List of the submodules
+    :rtype: builtin.list
+    """
+    module_info = loaded_module_info(module_name)
+    module_list = []
+    try:
+        submodules = module_info["submodules"]
+    except KeyError:
+        LOG.info("Module %s is not loaded", module_name)
+    else:
+        module_list = submodules
+        for module in submodules:
+            module_list += get_submodules(module)
+    return data_structures.ordered_list_unique(module_list)
+
+
+def unload_module(module_name):
+    """
+    Removes a module. Handles dependencies. If even then it's not possible
+    to remove one of the modules, it will throw an error.CmdError exception.
+
+    :param module_name: Name of the module we want to remove.
+    :type module_name: str
+    """
+    module_info = loaded_module_info(module_name)
+    try:
+        submodules = module_info['submodules']
+    except KeyError:
+        LOG.info("Module %s is already unloaded", module_name)
+    else:
+        for module in submodules:
+            unload_module(module)
+        module_info = loaded_module_info(module_name)
+        try:
+            module_used = module_info['used']
+        except KeyError:
+            LOG.info("Module %s is already unloaded", module_name)
+            return
+        if module_used != 0:
+            raise RuntimeError(f"Module {module_name} is still in use. "
+                               f"Can not unload it.")
+        process.system(f"/sbin/modprobe -r {module_name}")
+        LOG.info("Module %s unloaded", module_name)
+
+
+def module_is_loaded(module_name):
+    """
+    Is module loaded
+
+    :param module_name: Name of module to search for
+    :type module_name: str
+    :return: True if module is loaded
+    :rtype: bool
+    """
+    module_name = module_name.replace('-', '_')
+    return module_name in get_loaded_modules()
+
+
+def get_loaded_modules():
+    """
+    Gets list of loaded modules.
+    :return: List of loaded modules.
+    """
+    with open('/proc/modules', 'rb') as proc_modules:  # pylint: disable=W1514
+        return [astring.to_text(_.split(b' ', 1)[0]) for _ in proc_modules]
+
+
```

Comment:
> It is almost same from top to this line. It is not recommended that we have so many duplicated codes. We should find a solution to reuse avocado/utils/linux_modules.py, instead of simply copying its codes here.

Got it. I have updated this commit. Please review it. Thanks.